wal/wal.go

```diff
@@ -1042,71 +1037,44 @@ func validateRecord(typ recType, i int) error {
 // be a full record (recFull) if the record fits within the bounds of a single page.
 // Returns a byte slice of the record data read, the number of bytes read, and an error
 // if there's a non-zero byte in a page term record or the record checksum fails.
-// TODO(callum) the EOF errors we're returning from this function should theoretically
-// never happen, add a metric for them.
-func readRecord(buf []byte, header []byte, total int64) ([]byte, int, error) {
-	readIndex := 0
-	header[0] = buf[0]
-	readIndex++
-	total++
-
-	// The rest of this function is mostly from Reader.Next().
-	typ := recType(header[0])
-	// Gobble up zero bytes.
-	if typ == recPageTerm {
-		// We are pedantic and check whether the zeros are actually up to a page boundary.
-		// It's not strictly necessary but may catch sketchy state early.
-		k := pageSize - (total % pageSize)
-		if k == pageSize {
-			return nil, 1, nil // Initial 0 byte was last page byte.
+func readRecord(buf []byte, start, end int, header []byte) ([]byte, int, error) {
+	// Special case: for recPageTerm, check that are all zeros to end of page,
+	// consume them but don't return them.
+	if buf[start] == byte(recPageTerm) {
+		if end != pageSize {
```

Comment:
this probably works but is assuming we're always reading uncorrupted pages and so `recPageTerm` would always end at the end of the internal buffer

Reply:
`recPageTerm` should always end with the end of a page, and anything else is corrupt.  I think thats what this code honours?  If we see `recPageTerm` without a full page, we return EOF.  If we see `recPageTerm` and a full page, we check they are zero.  If they aren't, its corrupt, and we return the error.  Is that right?